programs/account-compression/src/instructions/nullify_leaves.rs

```diff
@@ -60,76 +84,69 @@ pub fn process_nullify_leaves<'a, 'b, 'c: 'info, 'info>(
 #[inline(never)]
 fn insert_nullifier(
     proofs: &[Vec<[u8; 32]>],
     change_log_indices: &[u64],
-    leaf: [u8; 32],
+    leaves: &[&[u8; 32]],
     indices: &[u64],
     ctx: &Context<'_, '_, '_, '_, NullifyLeaves<'_>>,
-    array_account: &mut std::cell::RefMut<'_, IndexedArrayAccount>,
-    leaves_queue_indices: &[u16],
 ) -> Result<Changelogs> {
-    let mut merkle_tree_account = ctx.accounts.merkle_tree.load_mut()?;
+    let mut merkle_tree = ctx.accounts.merkle_tree.load_mut()?;
 
-    if merkle_tree_account.associated_queue != ctx.accounts.indexed_array.key() {
+    if merkle_tree.associated_queue != ctx.accounts.indexed_array.key() {
         msg!(
             "Merkle tree and nullifier queue are not associated. Merkle tree associated nullifier queue {} != nullifier queue {}",
-            merkle_tree_account.associated_queue,
+            merkle_tree.associated_queue,
             ctx.accounts.indexed_array.key()
         );
         return Err(AccountCompressionErrorCode::InvalidIndexedArray.into());
     }
-    let loaded_merkle_tree = merkle_tree_account.load_merkle_tree_mut()?;
-    let allowed_proof_size = loaded_merkle_tree.height - loaded_merkle_tree.canopy_depth;
+    let merkle_tree = merkle_tree.load_merkle_tree_mut()?;
+    let allowed_proof_size = merkle_tree.height - merkle_tree.canopy_depth;
     if proofs[0].len() != allowed_proof_size {
         msg!(
             "Invalid Proof Length {} allowed height {} - canopy {} {}",
             proofs[0].len(),
-            loaded_merkle_tree.height,
-            loaded_merkle_tree.canopy_depth,
+            merkle_tree.height,
+            merkle_tree.canopy_depth,
             allowed_proof_size,
         );
         return Err(AccountCompressionErrorCode::InvalidMerkleProof.into());
     }
 
-    let mut bounded_vec = from_vec(proofs[0].as_slice())?;
-
-    let changelog_entries = loaded_merkle_tree
-        .update(
-            change_log_indices[0] as usize,
-            &leaf,
-            &ZERO_BYTES[0],
-            indices[0] as usize,
-            &mut bounded_vec,
-        )
-        .map_err(ProgramError::from)?;
-    let sequence_number = u64::try_from(loaded_merkle_tree.sequence_number)
+    let mut changelog_entries: Vec<ChangelogEntry26> = Vec::with_capacity(leaves.len());
+    for (i, leaf) in leaves.iter().enumerate() {
+        let mut proof = from_vec(proofs[i].as_slice()).map_err(ProgramError::from)?;
+        changelog_entries.push(
+            merkle_tree
+                .update(
+                    change_log_indices[i] as usize,
+                    leaf,
+                    &ZERO_BYTES[i],
+                    indices[i] as usize,
+                    &mut proof,
+                )
+                .map_err(ProgramError::from)?,
+        );
+    }
+
+    let sequence_number = u64::try_from(merkle_tree.sequence_number)
         .map_err(|_| AccountCompressionErrorCode::IntegerOverflow)?;
-    // TODO: replace with root history sequence number
```

Comment:
this TODO should remain and be moved to the line which executes ``mark_with_sequence_number``